crates/biome_js_analyze/src/analyzers/nursery/no_duplicate_test_hooks.rs

```diff
@@ -0,0 +1,221 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, AddVisitor, Phases, QueryMatch, Queryable, Rule,
+    RuleDiagnostic, ServiceBag, Visitor, VisitorContext,
+};
+use biome_console::markup;
+use biome_js_syntax::{AnyJsExpression, JsCallExpression, JsLanguage, JsSyntaxToken, TextRange};
+use biome_rowan::{AstNode, Language, SyntaxNode, WalkEvent};
+
+declare_rule! {
+    /// A `describe` block should not contain duplicate hooks.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// describe('foo', () => {
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   test('foo_test', () => {
+    ///    // some test
+    ///   });
+    /// });
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// describe('foo', () => {
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   test('foo_test', () => {
+    ///     afterAll(() => {
+    ///       // some teardown
+    ///     });
+    ///    afterAll(() => {
+    ///      // some teardown
+    ///    });
+    ///   });
+    /// });
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// describe('foo', () => {
+    ///   beforeEach(() => {
+    ///     // some setup
+    ///   });
+    ///   test('foo_test', () => {
+    ///     // some test
+    ///   });
+    /// });
+    /// ```
+    ///
+    pub(crate) NoDuplicateTestHooks {
+        version: "next",
+        name: "noDuplicateTestHooks",
+        recommended: true,
+    }
+}
+
+#[derive(Debug, Default)]
+struct HooksContext {
+    before_all: usize,
+    before_each: usize,
+    after_all: usize,
+    after_each: usize,
+}
+
+#[derive(Default)]
+struct DuplicateHooksVisitor {
+    stack: Vec<HooksContext>,
+}
+
+impl Visitor for DuplicateHooksVisitor {
+    type Language = JsLanguage;
+
+    fn visit(
+        &mut self,
+        event: &WalkEvent<SyntaxNode<Self::Language>>,
+        mut ctx: VisitorContext<Self::Language>,
+    ) {
+        match event {
+            WalkEvent::Enter(node) => {
+                // When the visitor enters a function node, push a new entry on the stack
+                if let Some(node) = JsCallExpression::cast_ref(node) {
+                    if let Ok(callee) = node.callee() {
+                        if callee.contains_a_test_pattern().is_ok() {
```

Comment:
We should check `if callee.contains_a_test_pattern() == Ok(true)`. The `SyntaxResult` is present only in case there's broken syntax, but if the syntax **is not broken**` we still could get a `false`